tests/generators/generators.js

```diff
@@ -0,0 +1,107 @@
+function *stmt_yield(): Generator<number, void, void> {
+  yield 0; // ok
+  yield ""; // error: string ~> number
+}
+
+function *stmt_next(): Generator<void, void, number> {
+  (yield undefined : number); // ok
+  (yield undefined : string); // error: number ~> string
+}
+
+function *stmt_return_ok(): Generator<void, number, void> {
+  return 0; // ok
+}
+
+function *stmt_return_err(): Generator<void, number, void> {
+  return ""; // error: string ~> number
+}
+
+function *infer_stmt() {
+  var x: boolean = yield 0;
+  return "";
+}
+for (var x of infer_stmt()) { (x : string) } // error: number ~> string
+var infer_stmt_next = infer_stmt().next(0).value; // error: number ~> boolean
+if (typeof infer_stmt_next === "undefined") {
+} else if (typeof infer_stmt_next === "number") {
+} else {
+  (infer_stmt_next : boolean) // error: string ~> boolean
+}
+
+declare function *decl(): Generator<number, string, boolean>;
+for (var x of decl()) { (x : string) } // error: number ~> string
+var decl_next = decl().next(0).value; // error: number ~> boolean
+if (typeof decl_next === "undefined") {
+} else if (typeof decl_next === "number") {
+} else {
+  (decl_next : boolean) // error: string ~> boolean
+}
+
+function *widen_next() {
+  var x = yield 0;
+  if (typeof x === "number") {
+  } else if (typeof x === "boolean") {
+  } else {
+    (x : string) // ok, sherlock
+  }
+}
+widen_next().next(0)
+widen_next().next("")
+widen_next().next(true)
+
+function *widen_yield() {
+  yield 0;
+  yield "";
+  yield true;
+}
+for (var x of widen_yield()) {
+  if (typeof x === "number") {
+  } else if (typeof x === "boolean") {
+  } else {
+    (x : string) // ok, sherlock
+  }
+}
+
+function *delegate_next_generator() {
+  function *inner() {
+    var x: number = yield undefined; // error: string ~> number
+  }
+  yield *inner();
+}
+delegate_next_generator().next("");
+
+function *delegate_yield_generator() {
+  function *inner() {
+    yield "";
+  }
+
+  yield *inner();
+}
+for (var x of delegate_yield_generator()) {
+  (x : number) // error: string ~> number
+}
+
+function *delegate_return_generator() {
+  function *inner() {
+    return "";
+  }
+
+  var x: number = yield *inner() // TODO: error: string ~> number
+}
+
+// only generators can make use of a value passed to next
+function *delegate_next_iterable(xs: Array<number>) {
+  yield *xs;
+}
+delegate_next_iterable([]).next(""); // TODO: error: string ~> void
```

Comment:
OK, so it turns out that this error message is actually caused by the intended failure, but the message is incredibly unhelpful because I was "too clever" in how I implemented the type checking here.

For reference, the `$yieldDelegate` declaration is:

```
declare var $yieldDelegate: (<R,N>(g: Generator<*,R,N>, n:N) => R) & ((i: Iterable<*>, n:void) => void);
```

Then the code flows the `yield*` expression's argument, and the parent generator's `next` tvar as arguments, and returns the return value. I'm using this construction because I don't know any other way to handle both Generators and Iterables like this. I picked up the trick from @bhosmer and @nmote's async/await changes.

So in this case, since the `yield*` expression's argument is an Iterable, we can pretend the declaration is the simpler `(i: Iterable<*>, n:void) => void`, into which we flow `Array<number>` (from `xs`) and `string` (from the `""` argument to `next()`), respectively.

Since the argument to the parameter `n` is non-`void`, flow complains, like I wanted it to. However, the error message blames the entire intersection type, not the specific flow `string ~> void`.

I actually already have a (possibly) related issue open for this at #341.

Any idea how I can either a) rewrite the type checking to avoid the `$yieldDelegate` function declaration trick of b) improve the messaging around intersection types to provide a more helpful error message?